Rationalize consistency checks on the number of columns

flomnes opened the pull request:
# Context
This is a preliminary work to place `maxHydroPumping` & `maxHydroGen` under sc-builder prefix "h", and remove prefix "hgp" introduced in CR23 (see #2068).

This should also fix #2069.

# Description
- Add homogeneity in the checks performed on the number of columns for
   - Links (direct & indirect) (prefix "ntc)
   - Hydro (ror, storage & mingen) (prefix "h")
   - Thermal (prefix "t", not useful for now, but we may use this mechanism for fuelCost & co2Cost, to replace the custom `checkFuelCostColumnNumber` and `checkCO2CostColumnNumber` functions)
- Declare TSNumbers before series. Otherwise we get a segfault since data members are initialized in the order where they appear in the class declaration.
- Remove `EqualizeGenerationTSsizes` mechanism for `{ror, storage}` and `{ror, mingen}`. If only one column is present, use it, which is more or less equivalent to the column duplication made by `EqualizeGenerationTSsizes`.

# TODO
- [x] Add unit tests, especially to check that errors are emitted when they should.

# Related work
[Improve logs](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2080)

flomnes committed:
Add mechanism to check TS consistency

flomnes committed:
Fix mingen reset

flomnes committed:
Fix multiple segfaults

flomnes committed:
Fix test segfault

flomnes committed:
Fix regression (wrong number of TS)

flomnes commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -157,11 +157,15 @@ void DataSeriesHydro::reset()
 {
     storage.reset(nbSeries, DAYS_PER_YEAR);
     ror.reset(nbSeries, HOURS_PER_YEAR);
-    mingen.reset(nbSeries, HOURS_PER_YEAR);
```
Fix #2069

flomnes committed:
Add unit tests

flomnes committed:
Simplify

flomnes committed:
Add unit test for single column time-series

flomnes committed:
fix loop

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -682,6 +682,39 @@ bool TimeSeriesNumbers::checkAllElementsIdenticalOrOne(std::vector<uint> w)
     return std::adjacent_find(w.begin(), first_one, std::not_equal_to<uint>()) == first_one;
 }
 
+bool TimeSeriesNumbers::CheckNumberOfColumns(const Study& study)
```
`bool TimeSeriesNumbers::CheckNumberOfColumns(const Study& study)` : 
argument **study** used only for `study.areas`. We should now the arg type to area list.

flomnes replied:
This has been done in PR #2080, see https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2080/files#diff-1bdb2c0970ab047ae53d0bcb2f829af2f797fac087c50d5f69ef08e9bbe217e1

flomnes replied:
DONE and MERGED

sylvlecl commented on src/libs/antares/series/include/antares/series/series.h:
```diff
@@ -35,10 +35,17 @@ namespace Antares::Data
 {
 public:
-    using numbers = Matrix<uint32_t>;
+    struct Numbers : public Matrix<uint32_t> // TODO[FOM] remove inheritance, it was added mostly to avoid breaking existing code
```
Indeed, would be better to use composition here with a class encapsulating both TS numbers and registered timeseries.
Do you prefer to postpone it to later work?

flomnes replied:
Yes I'd rather do it later

flomnes replied:
@sylvlecl done here #2085

sylvlecl commented on src/libs/antares/series/series.cpp:
```diff
@@ -31,9 +33,31 @@ using namespace Yuni;
+
+TimeSeries::TimeSeries(Numbers& tsNumbers) : timeseriesNumbers(tsNumbers)
+{
+    tsNumbers.registerSeries(this);
```
We should probably forbid copying of `TimeSeries`, to make sure that this is not called multiple times for the same timeseries.

flomnes replied:
I'd rather prevent copies, but it seems that `AreaScratchpad` must be copyable (used in std::vector), so build fails if I
- Delete copy/move constructor and operator
- Delete copy, keep move constructor and operator.

sylvlecl commented on src/libs/antares/series/series.cpp:
```diff
@@ -31,9 +33,31 @@ using namespace Yuni;
+// TODO[FOM] Code duplication
+static bool checkAllElementsIdenticalOrOne(std::vector<uint> w)
+{
+    auto first_one = std::remove(w.begin(), w.end(), 1); // Reject all 1 to the end
```
Not important but note that it's not the most efficient way of implementing this functions logic: without copying and moving anything, we could just register the first value different from 1, continue, and stop as soon as we encounter another one.

flomnes replied:
Yes, we had issues in the past with this function but you're right. We have a lot of UTs already so we could change implementation

sylvlecl commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -682,6 +682,39 @@ bool TimeSeriesNumbers::checkAllElementsIdenticalOrOne(std::vector<uint> w)
 
+bool TimeSeriesNumbers::CheckNumberOfColumns(const Study& study)
+{
+    std::map<const Antares::Data::TimeSeries::Numbers*, std::string> toCheck;
```
Indexing on a pointer should be avoided: the key order will vary from one execution to another, leading to different orders in execution and messages.
It's better to have a vector of pairs instead (or small structs for naming stuff), because we don't care about finding based on the pointer.

flomnes replied:
I agree, this is **bad**

flomnes replied:
DONE

sylvlecl commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -682,6 +682,39 @@ bool TimeSeriesNumbers::checkAllElementsIdenticalOrOne(std::vector<uint> w)
     return std::adjacent_find(w.begin(), first_one, std::not_equal_to<uint>()) == first_one;
 }
 
+bool TimeSeriesNumbers::CheckNumberOfColumns(const Study& study)
```
Having this method which validates all timeseries with the same logic is a lot better than before I think.

A thought for later though: when adding new timeseries numbers, it will still be quite easy to miss this check. Having a per-object `validate` method that would be called at the start of a computation could make it more "visible", so that we don't miss this validation step. This is linked to the broader validation topic we have already started to discuss.

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -682,6 +682,39 @@ bool TimeSeriesNumbers::checkAllElementsIdenticalOrOne(std::vector<uint> w)
+bool TimeSeriesNumbers::CheckNumberOfColumns(const Study& study)
+{
+    std::map<const Antares::Data::TimeSeries::Numbers*, std::string> toCheck;
+    for (const auto& [_, area] : study.areas)
```
I can understand why you check TS number of columns for **links** and **hydro**.
But why **thermal** clusters ?

flomnes replied:
It can be safely removed without any effect.

flomnes replied:
DONE

sylvlecl reviewed (changes requested):
Minor technical requests

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -682,6 +682,39 @@ bool TimeSeriesNumbers::checkAllElementsIdenticalOrOne(std::vector<uint> w)
+{
+    std::map<const Antares::Data::TimeSeries::Numbers*, std::string> toCheck;
+    for (const auto& [_, area] : study.areas)
+    {
```
This function has 2 parts because it seems you try to avoid duplication about logging.
But it looks too complicated and difficult to read to me.
We could only have one block for **links**  and one for **hydro**, and avoid the current last block (I skip **thermal** because of previous comment).

flomnes replied:
I will remove thermal. I can split up the for loop if needed.

guilpier-code commented on src/libs/antares/series/include/antares/series/series.h:
```diff
@@ -35,10 +35,17 @@ namespace Antares::Data
 {
 public:
-    using numbers = Matrix<uint32_t>;
+    struct Numbers : public Matrix<uint32_t> // TODO[FOM] remove inheritance, it was added mostly to avoid breaking existing code
```
**TimeSeries** should know nothing about TS numbers, I think class **TimeSeries** should be left alone.

On the other hand, class **Numbers** (to be renamed into **TSnumbers**) should be the one who knows about **TimeSeries**, because it checks some **TimeSeries** nmuber of columns consistency.

Instead, current classes **TimeSeries** and **TimeSeries** have a mutual dependency.

So, to me : 
- This class **Numbers** should be renamed into **TSnumbers**, and moved out from class **TimeSeries** (probably in a separate source file).
- We should reverse the registration process : make **Numbers** register **TimeSeries**, and not (as it is done now) make **TimeSeries** allow **Numbers** to register them (which is more complicated). 

For example, `DataSeriesHydro::DataSeriesHydro()` should make `timeseriesNumbers` register `ror`, `storage `and `mingen` : 

```C++
DataSeriesHydro::DataSeriesHydro() :
    timeseriesNumbers.register(ror, "ror"),
    timeseriesNumbers.register(storage, "storage"),
    timeseriesNumbers.register(mingen, "mingen")
{
    ...
}
```

flomnes replied:
Yes, will do

flomnes replied:
DONE

guilpier-code commented on src/solver/simulation/include/antares/solver/simulation/solver.hxx:
```diff
@@ -339,6 +339,11 @@ void ISimulation<ImplementationType>::run()
         // for a single simulation
         study.resizeAllTimeseriesNumbers(1 + study.runtime->rangeLimits.year[Data::rangeEnd]);
         // Now, we will prepare the time-series numbers
+        if (not TimeSeriesNumbers::CheckNumberOfColumns(study))
```
Good : resizing due to TS generation was already made at this stage

flomnes replied:
Yes !

guilpier-code commented on src/libs/antares/study/include/antares/study/parts/hydro/series.h:
```diff
@@ -144,10 +144,6 @@ class DataSeriesHydro
-    Matrix<uint32_t> timeseriesNumbersHydroMaxPower;
-
     // Equalizing max generation and max pumping numbers of TS's    
     void EqualizeMaxPowerTSsizes(Area& area);
```
We still have an **Equalize** function for max power pumping and generation.
We'll have to remove it at some point.

flomnes replied:
See #2068

flomnes committed:
Add logs to checks on number of columns, refactor (#2080)

Add more context to allow the user to correct invalid data.

Extract of logs after this PR
```
[2024-05-03 13:43:44][solver][error] Inconsistent number of columns for link fr / gb (direct-capacity: 3, indirect-capacity: 2)
[2024-05-03 13:43:44][solver][error] Please check that all series have the same number of columns or 1 column, or a combination of 1 column and the same number of columns
```

Also, small refactor to split the implementation of function
`TimeSeriesNumbers::CheckNumberOfColumns` in 2.

Using a `std::map` induces a slight change of behavior : unique keys. TS
with the same key will be erased. This is mostly good, for example if
multiple `AreaScratchpad` are created for a given area (parallel
simulations).

flomnes committed:
PR review

flomnes committed:
Use composition for TimeSeriesNumbers (instead of inheritance) (#2085)

Also, use `timeSeriesNumbers[y]` instead of `timeSeriesNumbers[0][y]`.

flomnes committed:
Apply new formatting

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/check-consistency-ts-numbers

flomnes committed:
Reduce captured context
